### PR TITLE
storage: finalize dropped shards take 2

### PR DESCRIPTION
**Pull request by sploiselle**
Re-do of #17327 that should take #18378 into account.

@aljoscha @danhhz @bkirwi -- does this approach look right/sane to you all?

### Motivation

  * This PR adds a known-desirable feature. Discussed in #17327
  * This PR fixes a recognized bug.  #17327 was deficient in that it caused #18378

### Tips for reviewer

Putting this in draft primarily as a talking point; maybe there are other viable approaches we should consider.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by aljoscha** [comment]
What's the high-level diff from the previous incarnation (#17327)?

**Comment by sploiselle** [comment]
@aljoscha We previously:
- Sealed the shard from further writes, irrespective of any since.
- Processed each shard at most once, meaning we had to close it then or never.

The approach in this PR instead:
- Only seals the shard from writes if the global since is empty.
- Only clears the shards from the finalization WAL if the global since is empty, meaning we will periodically cycle through all of the shards we might want to close until we are more confident that no one else is using them

**Comment by danhhz** [comment]
yeah, that approach sounds sane. worst case, the leased readers will eventually be expired. how does the critical since handle play into this?

**Comment by aljoscha** [comment]
> * Only clears the shards from the finalization WAL if the global since is empty

That bit we already did in the previous PR, right? It's still good here, just double-checking.

**Comment by sploiselle** [comment]
@aljoscha Unfortunately, this is wrong on `main`; we remove shards from the WAL once they're passed through `finalize_shards` irrespective of their since. Fortunately, nothing currently uses this code.

@danhhz We were not expiring these before; is https://github.com/MaterializeInc/materialize/pull/18461 the proper way to handle trying to fully drop and clean up the critical since handle? Or is the suggestion to _not_ do it as in #18461 and instead do it in this new WAL processing code?

**Comment by danhhz** [comment]
> @danhhz We were not expiring these before; is https://github.com/MaterializeInc/materialize/pull/18461 the proper way to handle trying to fully drop and clean up the critical since handle? Or is the suggestion to not do it as in https://github.com/MaterializeInc/materialize/pull/18461 and instead do it in this new WAL processing code?

not suggesting anything in particular! I'm not familiar enough with the storage controller to have confidant opinions on where this should go. what I'm trying to get at is just making sure that no matter what happens, we _always_ eventually get overall since to empty, so things don't get stuck in the WAL forever. if you're looking for opinions on how that should happen, maybe @aljoscha has one?

**Comment by aljoscha** [comment]
> @aljoscha Unfortunately, this is wrong on `main`; we remove shards from the WAL once they're passed through `finalize_shards` irrespective of their since. Fortunately, nothing currently uses this code.

Now I'm confused, I thought this change from the previous incarnation of the change made it so we only remove them once the since is `[]`

**Comment by sploiselle** [comment]
@aljoscha lmk if you want to hop on a call; this is correct:

> we only remove <shards from the WAL in this PR> once the since is `[]`

Not sure what the miscommunication in my previous statement was, and won't hazard a guess lest I risk further confusion :D

**Comment by sploiselle** [comment]
To summarize Dan's comment https://github.com/MaterializeInc/materialize/pull/18449#issuecomment-1487610181, the concern is if the shards in the WAL will ever make progress toward having an empty since or if the critical read handle's hold will "wedge" them at some point in the past.

As far as the current uses of the function are concerned, we are only introducing shards into the WAL that are in the "happy path" that stems from the collections being dropped, which does downgrade the shard's since to `[]` in the `PersistReadWorker`.

In the future, if we want to add shards to this WAL without going through the `drop_source` code path, we'll need to ensure their critical read handles are downgraded.

**Comment by sploiselle** [comment]
@philip-stoev This is a change that previously needed to be reverted due to failing nightlies. I've validated that this does, in fact, pass those previously failing tests. You mind checking/OKing this? Ty

File: src/storage-client/src/controller.rs
Comment (aljoscha) [inline src/storage-client/src/controller.rs:2730]: This could fail and we could handle it gracefully by just retrying next time `finalize_shards` is called, meaning we'd just not panic here because the rest already works?

File: src/storage-client/src/controller/command_wals.rs
Comment (aljoscha) [inline src/storage-client/src/controller/command_wals.rs:142]: Why don't we need this anymore? I'm probably just forgetting sth, it has been a while ... 🙈

File: src/storage-client/src/controller.rs
Comment (aljoscha) [inline src/storage-client/src/controller.rs:2055]: Double-checking a problem we've had in the past: do we get `DroppedIds` messages for all the subsources as well, or do _we_ have to make sure to register them here?

~~Also, I think that TODO is still valid? We still don't delete them from our collection state, do we?~~ Ah no, `finalize_shards`, does that right?

**Review by aljoscha — APPROVED** [review]
Approving now, because the overall approach and code looks good! Please merge once you feel you've answered or thought about my comments enough! 😅 And once you've heard back from Philip.

File: src/storage-client/src/controller.rs
Comment (sploiselle) [reply src/storage-client/src/controller.rs:2055]: 1. Yes, we do get the subsource IDs here, as well.
2. This TODO should not have been dropped here because the do not remove the `GlobalId` from `self.state.collections`; all of this code got taken from a branch that _does_ do this work, which was why it got removed. Sorry for the confusion here.

**Comment by sploiselle** [comment]
@philip-stoev @MaterializeInc/qa -- the last attempt at implementing this feature caused #18378; I've tested locally to ensure that this time we don't run into the same issue, but can I get a blessing on this approach? Maybe a nightly run?

**Pull request merged**
